File: 106.TreapAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.util.*;

// Node structure for the Treap
class Node {
    int key;           // key maintains BST property
    int priority;      // priority maintains Heap property
    Node left, right;  // left and right child pointers

    Node(int key) {
        this.key = key;
        this.priority = new Random().nextInt(100); // assign random priority
        this.left = null;
        this.right = null;
    }
}

public class TreapExample {

    // Right rotation to fix heap property violation
    static Node rotateRight(Node y) {
        Node x = y.left;          // store left child
        Node T = x.right;         // store right subtree of x
        x.right = y;              // perform rotation
        y.left = T;               // attach T as left subtree of y
        return x;                 // new root after rotation
    }

    // Left rotation to fix heap property violation
    static Node rotateLeft(Node x) {
        Node y = x.right;         // store right child
        Node T = y.left;          // store left subtree of y
        y.left = x;               // perform rotation
        x.right = T;              // attach T as right subtree of x
        return y;                 // new root after rotation
    }

    // Insert a key into the Treap
    static Node insert(Node root, int key) {
        // base case: if tree is empty, return new node
        if (root == null)
            return new Node(key);

        // insert key in BST manner
        if (key < root.key) {
            root.left = insert(root.left, key);

            // fix heap property if violated
            if (root.left.priority > root.priority)
                root = rotateRight(root);
        } 
        else if (key > root.key) {
            root.right = insert(root.right, key);

            if (root.right.priority > root.priority)
                root = rotateLeft(root);
        }

        // return the unchanged or rotated root
        return root;
    }

    // Search for a key in Treap (like normal BST)
    static boolean search(Node root, int key) {
        if (root == null)
            return false;
        if (root.key == key)
            return true;
        else if (key < root.key)
            return search(root.left, key);
        else
            return search(root.right, key);
    }

    // Inorder traversal to display the Treap
    static void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.print("(Key=" + root.key + ", Priority=" + root.priority + ") ");
            inorder(root.right);
        }
    }

    public static void main(String[] args) {
        Node root = null; // start with empty Treap

        // insert elements into the Treap
        int[] keys = {50, 30, 20, 40, 70, 60, 80};
        for (int k : keys)
            root = insert(root, k);

        // print Treap inorder (shows BST property)
        System.out.println("Inorder traversal of Treap (Key, Priority):");
        inorder(root);

        // search operations
        System.out.println("\nSearch for 40: " + search(root, 40));  // expected true
        System.out.println("Search for 90: " + search(root, 90));    // expected false
    }
}

File: 106.TreapAlgo.py
# A Treap (Tree + Heap) is a randomized balanced binary search tree (BST) that combines
# the properties of both a Binary Search Tree (BST) and a Heap.
# Each node has a key (satisfying BST property) and a priority
# (satisfying Heap property), ensuring balance through randomization.

# Explanation

# A Treap maintains two conditions:

# BST Property:
# For every node, all keys in the left subtree are smaller, and all keys in the right subtree are larger.

# Heap Property:
# For every node, its priority is higher (or lower, depending on convention) than that of its children.

# When inserting:

# Insert the node as in a normal BST by key.

# Then rotate to fix the heap property if violated.

# When deleting:

# Rotate down the node to be deleted until it becomes a leaf, then remove it.

# Because priorities are assigned randomly, Treap remains balanced on average
# with an expected height of O(log n).

import random


class Node:
    def __init__(self, key):
        self.key = key
        self.priority = random.randint(1, 100)
        self.left = None
        self.right = None


def rotate_right(y):
    x = y.left
    T = x.right
    x.right = y
    y.left = T
    return x


def rotate_left(x):
    y = x.right
    T = y.left
    y.left = x
    x.right = T
    return y


def insert(root, key):
    if root is None:
        return Node(key)

    if key < root.key:
        root.left = insert(root.left, key)
        if root.left.priority > root.priority:
            root = rotate_right(root)

    elif key > root.key:
        root.right = insert(root.right, key)
        if root.right.priority > root.priority:
            root = rotate_left(root)

    return root


def search(root, key):
    if root is None:
        return False
    if root.key == key:
        return True
    elif key < root.key:
        return search(root.left, key)
    else:
        return search(root.right, key)


def inorder(root):
    if root:
        inorder(root.left)
        print(f"(key = {root.key}, Priority={root.priority})", end=" ")
        inorder(root.right)


root = None
# Insert keys into Treap
for k in [50, 30, 20, 40, 70, 60, 80]:
    root = insert(root, k)

print("Inorder traversal of Treap (Key, Priority):")
inorder(root)
print("\nSearch for 40:", search(root, 40))
print("Search for 90:", search(root, 90))
